Rough draft for rep2010: service introspection

ihasdapie opened the pull request:
Here's a rough draft of the service introspection REP with all the sections filled in somewhat

ihasdapie committed:
fix swapped service/client in msg definition comments

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

ihasdapie committed:
rough draft

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

deepanshubansal01 commented on rep-2010.rst:
```diff
@@ -231,47 +257,77 @@ These topics are vulnerable to undesired actors listening in on service communic
+Using ROS Parameters for configuration
+--------------------------------------
+ROS parameters provide a well-known and supported way of interacting with node-level settings at various points in a node's lifecycle.
+An user may want to avoid the overhead of publishing service event messages during normal operation,
```
minor thing but **A user** instead of **An user**

ihasdapie replied:
Right. English. Vowel sound

deepanshubansal01 commented on rep-2010.rst:
```diff
@@ -231,47 +257,77 @@ These topics are vulnerable to undesired actors listening in on service communic
 
-TODO
+Duplicating request or response content will be optional as service response messages can be *very* large and flooding the ROS network with up to four copies of the same message can introduce undesirable load and latency.
+By default content will be included on ``REQUEST_SENT`` and ``RESPONSE_SENT``. 
```
**REQUEST_SENT** can be renamed to **CLIENT_SENT** and **RESPONSE_SENT** to **SERVICE_SENT** according to new naming convention.

deepanshubansal01 commented on rep-2010.rst:
```diff
@@ -220,7 +219,34 @@ And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
+         SERVICE_NAME/service_events_enabled: bool
+         
+         # Indicates if request content will be included in CLIENT_SENT RequestEvent messages
+         SERVICE_NAME/_enable_client_sent_content: bool
+
+         # Indicates if request content will be included in SERVICE_RECEIVED RequestEvent messages
+         SERVICE_NAME/_enable_service_receive_content: bool
+
+         # Indicates if response content will be included in SERVICE_SENT RequestEvent messages
+         SERVICE_NAME/_enable_service_sent_content: bool
+
+         # Indicates if response content will be included in CLIENT_RECEIVED RequestEvent messages
+         SERVICE_NAME/_enable_client_receive_content: bool
```
Out of curiosity why aren't we going with the enum approach for `_enable_content`. I remember their were some reasons for not using enum's I guess but not sure. 

We could do something like:
`SERVICE_NAME/_enable_content enum_value`

ihasdapie replied:
There would be 2^4 possible enums which doesn't sound ideal. This is also a ROS parameter and we would want it to be settable via YAML files

jacobperron replied:
I think the main problem is that it's difficult to reference enums when setting parameters from the CLI or YAML file. There is not "enum" type for ROS parameters.

jacobperron replied:
I've added a note about this to the rationale section [6419e51](https://github.com/jacobperron/rep/pull/2/commits/6419e51e61766535b7b0c5a62a960850d6fcabd2)

ihasdapie committed:
follow through with renaming

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

ihasdapie committed:
an user -> a user (vowel sound...)

Signed-off-by: Brian Chen <brian.chen@openrobotics.org>

jacobperron commented on rep-2010.rst:
```diff
@@ -220,7 +219,34 @@ And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
-TODO: Describe what configuration options there are and how we plan to expose them (e.g. with ROS parameters and/or environment variables).
+Configuration of service introspection features will be done through ROS parameters.
+
+.. code-block:: yaml
```
I think it would be better to just list out parameters and their descriptions in rst, and we can conclude the section with a concrete example in YAML.

jacobperron replied:
See [83f0841](https://github.com/jacobperron/rep/pull/2/commits/83f084112210ca4a6bb279f11ef93a5311b783db)

jacobperron commented on rep-2010.rst:
```diff
@@ -220,7 +219,34 @@ And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
 -------------
 
-TODO: Describe what configuration options there are and how we plan to expose them (e.g. with ROS parameters and/or environment variables).
+Configuration of service introspection features will be done through ROS parameters.
```
It's probably also worth noting that we want these to be runtime configurable.

jacobperron replied:
[83f0841](https://github.com/jacobperron/rep/pull/2/commits/83f084112210ca4a6bb279f11ef93a5311b783db)

jacobperron commented on rep-2010.rst:
```diff
@@ -231,47 +257,77 @@ These topics are vulnerable to undesired actors listening in on service communic
+----------------
+
+ROS 2 parameters are built on top of ROS 2 service/clients and already publish out to ``/parameter_events`` on parameter changes [11]_.
+As parameters can be set directly by the host as well as through the corresponding parameter service, the service introspection proposed by this REP would fail to capture a subset of parameter events.
```
```suggestion
As parameters can be set directly by the host node as well as through the corresponding parameter service, the service introspection proposed by this REP would fail to capture a subset of parameter events.
```

jacobperron replied:
I've updated this section (and the spec) regarding parameter services ([d51c900](https://github.com/jacobperron/rep/pull/2/commits/d51c900dbe0a478133bd98c7f1981500fae3002d)).
I'm hesitant to propose an additional configuration option specifically for them at this point in time. Let's wait for more feedback from the community and to see how the prototype turns out.

jacobperron commented on rep-2010.rst:
```diff
@@ -231,47 +257,77 @@ These topics are vulnerable to undesired actors listening in on service communic
+This is to best balance avoiding introducing additional topic surface area and complexity of consuming emitted introspection data. 
+Grouping events by request or response rather than by client or subscriber would also ease request/response tracing work.
+
+.. Concern: our enable flags are enable by serv/client but our messages are grouped via req/response. Could be confusing?
```
Maybe? Let's see if this is a concern raised by other reviewers.

jacobperron commented on rep-2010.rst:
```diff
@@ -231,47 +257,77 @@ These topics are vulnerable to undesired actors listening in on service communic
-TODO
+Service introspection is a new feature which is not planned to impact any existing logic. 
+It is also an opt-in feature so there will be no additional incurred overhead by default.
+As always, code written to take advantage of this new feature will not be backwards compatible (at least regarding this feature) with versions of ROS 2 that do not have it implemented.
```
I think this is a given and not worth mentioning.

jacobperron replied:
Minor refactor of this section: [308f7c5](https://github.com/jacobperron/rep/pull/2/commits/308f7c58622f97be29d7fab0f0831c4735058531)

jacobperron committed:
Minor formatting and name fixes

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented on rep-2010.rst:
```diff
@@ -220,7 +219,34 @@ And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
+      ros__parameters:
+         
+         # Indicates if CLIENT_SENT and CLIENT_RECEIVED events will be published
+         SERVICE_NAME/client_events_enabled: bool
```
I'm not sure if we want to include the service name in the parameter names, then we'll end up with N * Sn * 6 parameters, where N is number of nodes and Sn is number of services for each node. Though this gives the user fine grain control over which services provided by a node will emit events, I'm not sure if it's worth the added complexity. I imagine the more common use case for the feature will be to turn it on for all services for a node (or all nodes in the system). We could go so far as having a single parameter per node, but I think a good compromise are the existing 6 parameters that give the user some control over the number (and size of) event messages.

jacobperron replied:
See [83f0841](https://github.com/jacobperron/rep/pull/2/commits/83f084112210ca4a6bb279f11ef93a5311b783db)

ihasdapie replied:
@deepanshubansal01 and I had a discussion about this and figured it would be nice in the event that a node offers multiple large services. We can wait on some other feedback on this, though I agree that enabling/disabling for all services in a node should cover most cases.

jacobperron commented on rep-2010.rst:
```diff
@@ -220,7 +219,34 @@ And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
+         SERVICE_NAME/service_events_enabled: bool
+         
+         # Indicates if request content will be included in CLIENT_SENT RequestEvent messages
+         SERVICE_NAME/_enable_client_sent_content: bool
```
I'm not convinced that having the granularity to control both sent/received content is worth the additional complexity. IMO, one parameter for both client events and one parameter for both service events should be sufficient.

jacobperron replied:
[83f0841](https://github.com/jacobperron/rep/pull/2/commits/83f084112210ca4a6bb279f11ef93a5311b783db)

jacobperron commented on rep-2010.rst:
```diff
@@ -220,7 +219,34 @@ And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
+         SERVICE_NAME/_enable_client_receive_content: bool
+         
+         # Indicates if service introspection will be enabled for default services
+         _enable_default_service_introspection: bool 
```
I think this parameter becomes obsolete if we decide to omit the service name prefix from parameters.

jacobperron committed:
Refactor configuration section and rationale

* Reduce number of proposed parameters to 4 per node.
* Update rationale accordingly.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add note about using multiple bools instead of a enum

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update spec regarding parameter services

Noting that the behavior will be implementation defined for now.
I don't think we'll know how this will work until we start implementing the feature, and I'm hestiant the increase complexity by adding a special configuration option for parameter services at this point in time.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Minor wording refactor in backwards compatibility

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add back 'feature progress' section

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
@ihasdapie Thanks for adding all this content :+1:

I've done some refactoring and will merge this to prepare the REP for wider consumption. Feel free to comment here or on the upstream PR when it is up.
